src/python/pants/core/util_rules/external_tool.py

```diff
@@ -73,6 +97,15 @@ def my_rule(my_external_tool: MyExternalTool) -> Foo:
     default_version: str
     default_known_versions: List[str]
 
+    # our hard constraints required for this tool, to support the API we expect from it
+    version_constraints: Optional[str] = None
+    constraints: Optional[Requirement] = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.version_constraints:
+            self.constraints = Requirement.parse(f"{self.name}{self.version_constraints}")
```

Comment:
Rather than putting your check in `get_request`, I think that you can put it here: after calling `super()` you will be able to consume `self.options`, which means that you can arrange this so that `self.check_version_constraint` is callable here.

Reply:
Yes, that is much better to be in `__init__` because it's eager. Otherwise you will only lazily get an error when the tool is downloaded.

Reply:
Great. My attempt at making it eager ended up as a option validation, and became too clumsy. `__init__` it is :)